tools/kth_threshold.cpp

```diff
@@ -0,0 +1,204 @@
+#include <iostream>
+#include <optional>
+#include <unordered_set>
+
+#include "boost/algorithm/string/classification.hpp"
+#include "boost/algorithm/string/split.hpp"
+#include <boost/functional/hash.hpp>
+
+#include "mio/mmap.hpp"
+#include "spdlog/sinks/stdout_color_sinks.h"
+#include "spdlog/spdlog.h"
+
+#include "mappable/mapper.hpp"
+
+#include "app.hpp"
+#include "cursor/max_scored_cursor.hpp"
+#include "index_types.hpp"
+#include "io.hpp"
+#include "query/algorithm.hpp"
+#include "util/util.hpp"
+#include "wand_data_compressed.hpp"
+#include "wand_data_raw.hpp"
+
+#include "query/algorithm.hpp"
+#include "scorer/scorer.hpp"
+
+#include "CLI/CLI.hpp"
+
+using namespace pisa;
+
+template <typename IndexType, typename WandType>
+void kt_thresholds(
+    const std::string& index_filename,
+    const std::string& wand_data_filename,
+    const std::vector<Query>& queries,
+    std::string const& type,
+    ScorerParams const& scorer_params,
+    uint64_t k,
+    bool quantized,
+    std::optional<std::string> pairs_filename,
+    std::optional<std::string> triples_filename)
+{
+    IndexType index;
+    mio::mmap_source m(index_filename.c_str());
+    mapper::map(index, m);
+
+    WandType wdata;
+
+    auto scorer = scorer::from_params(scorer_params, wdata);
+
+    mio::mmap_source md;
+    std::error_code error;
+    md.map(wand_data_filename, error);
+    if (error) {
+        spdlog::error("error mapping file: {}, exiting...", error.message());
+        std::abort();
+    }
+    mapper::map(wdata, md, mapper::map_flags::warmup);
+
+    using Pair = std::set<uint32_t>;
+    std::unordered_set<Pair, boost::hash<Pair>> pairs_set;
+
+    using Triple = std::set<uint32_t>;
+    std::unordered_set<Triple, boost::hash<Triple>> triples_set;
+
+    auto to_int = [](std::vector<std::string> const& term_ids) {
```

Comment:
I don't really see any reason to create lambda over a free function, it doesn't capture anything and makes the main function longer and more difficult to decode.

But more importantly, why not just pass the line and split inside function? It's exactly the same in both cases. The `while` loop should be something like:

```cpp
while (std::getline(pin, line)) {
    pairs_set.insert(parse_tuple(line, 2)); // substitute 2 with 3 for triples
}
```
The reason why I'm passing 2 is that you want to make sure that you got the right number of terms, because if you pass a file that's four terms on each line, then nobody's gonna notice, there will be no error but unexpected behavior.

So below, you want to do:
```cpp
if (term_ids.size() != k) {
    throw std::runtime_error(fmt::format("Wrong number of terms in line: {} (expected {} but found {})", ...));
}
```